solver/src/liquidity/uniswap.rs

```diff
@@ -33,33 +37,12 @@ struct Inner {
     allowances: Mutex<HashMap<H160, U256>>,
 }
 
-impl UniswapLikeLiquidity {
-    pub fn new(
-        router: IUniswapLikeRouter,
-        pair_provider: Arc<dyn AmmPairProvider>,
-        gpv2_settlement: GPv2Settlement,
-        base_tokens: HashSet<H160>,
-        web3: Web3,
-    ) -> Self {
-        Self {
-            inner: Arc::new(Inner {
-                router,
-                gpv2_settlement,
-                allowances: Mutex::new(HashMap::new()),
-            }),
-            web3: web3.clone(),
-            pool_fetcher: PoolFetcher {
-                pair_provider,
-                web3,
-            },
-            base_tokens,
-        }
-    }
-
+#[async_trait::async_trait]
+impl BaselineLiquidity for UniswapLikeLiquidity {
     /// Given a list of offchain orders returns the list of AMM liquidity to be considered
-    pub async fn get_liquidity(
+    async fn get_liquidity(
         &self,
-        offchain_orders: impl Iterator<Item = &LimitOrder> + Send + Sync,
+        offchain_orders: &mut (dyn Iterator<Item = &LimitOrder> + Send + Sync),
```

Comment:
You need a mutable reference for `Iterator::next(&mut self)`. It doesn't mean that the collection is mutable, only that the iterator is (so that it can progress).

Reply:
I see. And the reason to change from `impl Iterator` to `&mut dyn Iterator` is so that it can be used in a trait?